### PR TITLE
Ruff linter + pre-commit integration

### PR DESCRIPTION
**Pull request by juanitorduz**
Closes https://github.com/pymc-devs/pymc-bart/issues/139

In addition, we move to https://pre-commit.ci/ to check the code style.

I took most of the code from the `pymc` repo ;)

### CODE REVIEW

**Comment by juanitorduz** [comment]
@aloctavodia I think one is ready for review. Please take a careful look because some changes were made automatically by Ruff. For more info on the options and config you can, of course, ask me ;) Or look into the pymc PR https://github.com/pymc-devs/pymc/pull/7091 where there were many useful discussions :D

**Comment by juanitorduz** [comment]
Hehe! I forgot to add `pylint`. I am on it 😅

**Comment by juanitorduz** [comment]
ok! All 🟢 !

File: requirements-dev.txt [hunk 1]
```diff
@@ -1,7 +1,4 @@
-black==22.3.0
 click==8.0.4
-mypy==1.3.0
```
Comment (aloctavodia) [inline requirements-dev.txt:3]: Is this right? thus Ruff also takes care of type hints?

**Review by aloctavodia — COMMENTED** [review]
LGTM. I think, we should probably use this  for other projects too!

File: requirements-dev.txt (hunk 1 above)
Comment (juanitorduz) [reply requirements-dev.txt:3]: No, but I am passing everything to the pre-commit hook: see https://results.pre-commit.ci/run/github/534281694/1706217149.DTa7osJdTE-SvRWjeQDG2Q It is still in

https://github.com/juanitorduz/pymc-bart/blob/18554ab35a4dbfce8bd74278fbffa2cb216c8d71/.pre-commit-config.yaml#L21-#L27

with the specific version

**Comment by aloctavodia** [comment]
After this is merged what is the contribution workflow? What do I need to run locally? I realized we should list the steps here https://github.com/pymc-devs/pymc-bart/blob/main/CONTRIBUTING.md @PabloGGaray

**Comment by juanitorduz** [comment]
> After this is merged what is the contribution workflow? What do I need to run locally? I realized we should list the steps here https://github.com/pymc-devs/pymc-bart/blob/main/CONTRIBUTING.md @PabloGGaray

True! I will add the steps on the PR. For reference: the pre-commit should take care of everything.

1. Install pre-commit

```bash
pip install pre-commit
```

2. Set up pre-commit

```bash
pre-commit install
```

3. Run the complete pre-commit hook to check specific files:

```bash
pre-commit run --files pymc_bart/tree.py
```

or all files:

```bash
pre-commit run --all-files
```

**Once you commit something the pre-commit hook will run all the checks**

You can skip this (for example when is WIP) by adding a flag (`-n` means no-verify)

```bash
git commit -m"my message" -n
```

One can, of course, install `ruff` in the Python environment to enable auto-format, but this is not necessary. The specific versions of` ruff` and `mypy` will be only specified in `.pre-commit-config.yaml`. It should be the only source of truth. 

(I will add this to the guidelines)

**Comment by aloctavodia** [comment]
Thanks a lot @juanitorduz!

**Pull request merged**